### PR TITLE
Add tests for User

### LINKED ISSUE #9: Add tests for user routes

### PR DESCRIPTION
**Pull request by sirkrypt0**

### CODE REVIEW

**Comment by frcroth** [comment]
@sirkrypt0 Perhaps you can remove the pull_request Hook in the CI in this PR to reduce the number of checks, as it seems unnecessary.

**Comment by sirkrypt0** [comment]
> @sirkrypt0 Perhaps you can remove the pull_request Hook in the CI in this PR to reduce the number of checks, as it seems unnecessary.

We could do something like the following, which should be sufficient imo. PRs should be opened as Draft as soon as work on a new feature has begun, which would make the `push` hook on other branches than the `main` obsolete.

```

on:
  push:
    branches: [ main ]
  pull_request:
    branches: [ main ]
```

**Comment by sirkrypt0** [comment]
> > @sirkrypt0 Perhaps you can remove the pull_request Hook in the CI in this PR to reduce the number of checks, as it seems unnecessary.
> 
> We could do something like the following, which should be sufficient imo. PRs should be opened as Draft as soon as work on a new feature has begun, which would make the `push` hook on other branches than the `main` obsolete.
> 
> ```
> 
> on:
>   push:
>     branches: [ main ]
>   pull_request:
>     branches: [ main ]
> ```

We might as well allow the CI on all pull requests, not only on main.

**Comment by frcroth** [comment]
Yes all PRs and push on main seems like a sensible setting.

**Comment by sirkrypt0** [comment]
@frcroth I think we can merge this already. I will add route/integration tests in another PR. This allows new MRs to already profit from the test helper and write tests for the model.

File: server/src/user/model.rs [hunk 1]
```diff
@@ -65,3 +65,114 @@ impl User {
         Ok(count)
     }
 }
+
```
Comment (frcroth) [inline server/src/user/model.rs:86]: I'm strongly for moving the tests to another file

File: server/src/user/model.rs [hunk 2]
```diff
@@ -65,3 +65,114 @@ impl User {
...
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test_helpers::*;
+
+    fn create_user_data() -> UserData {
+        UserData {
+            username: String::from("username"),
+            password: String::from("password"),
```
Comment (frcroth) [inline server/src/user/model.rs]: Can we use better standard data?
```suggestion
            username: String::from("testUser"),
            password: String::from("12345678"),
```

Or alternatively use random strings

File: server/src/user/model.rs [hunk 3]
```diff
@@ -65,3 +65,114 @@ impl User {
...
+    #[test]
+    fn destroy_returns_null_if_user_does_not_exist() {
+        let conn = connection();
+
+        let count = User::destroy(&conn, Uuid::new_v4()).unwrap();
+        assert_eq!(count, 0);
+    }
+
+    #[test]
+    fn destroy_returns_count_if_user_exists() {
```
Comment (frcroth) [inline server/src/user/model.rs]:
```suggestion
    fn destroy_returns_one_if_user_exists() {
```

**Review by frcroth — CHANGES_REQUESTED** [review]

File: server/src/user/model.rs (hunk 1 above)
Comment (sirkrypt0) [reply server/src/user/model.rs:86]: Actually, the book states:

>  You’ll put unit tests in the src directory in each file with the code that they’re testing. The convention is to create a module named tests in each file to contain the test functions and to annotate the module with cfg(test).
> https://doc.rust-lang.org/book/ch11-03-test-organization.html#unit-tests

Comment (sirkrypt0) [reply server/src/user/model.rs:86]: Integration tests will be placed into a `tests` directory

**Comment by frcroth** [comment]
Ok I see now that is rust convention to put the tests in the same file. I do not like it but I will accept it.

**Comment by sirkrypt0** [comment]
> Ok I see now that is rust convention to put the tests in the same file. I do not like it but I will accept it.

Actually I find it reasonable to have the tests in the same file. That way one can use the tests directly to understand what the code is doing and does not have to browse many files to find the tests.

**Review by frcroth — APPROVED** [review]
Requires some changes because the passwords are now saved as hash, other than that I approve

**Comment by sirkrypt0** [comment]
> Requires some changes because the passwords are now saved as hash, other than that I approve

I updated the tests and added a new one for `generate_password`. Please check if you agree with them.

**Review by frcroth — APPROVED** [review]

**Pull request merged**